src/thoas_decode.erl

```diff
@@ -1073,23 +1077,45 @@ string_decode_function(Options) ->
         copy -> fun binary:copy/1
     end.
 
-terminate(<<32/integer,Rest/bitstring>>, Input, Skip, Stack, StringDecode, Value) ->
-    terminate(Rest, Input, Skip + 1, Stack, StringDecode, Value);
-terminate(<<13/integer,Rest/bitstring>>, Input, Skip, Stack, StringDecode, Value) ->
-    terminate(Rest, Input, Skip + 1, Stack, StringDecode, Value);
-terminate(<<10/integer,Rest/bitstring>>, Input, Skip, Stack, StringDecode, Value) ->
-    terminate(Rest, Input, Skip + 1, Stack, StringDecode, Value);
-terminate(<<9/integer,Rest/bitstring>>, Input, Skip, Stack, StringDecode, Value) ->
-    terminate(Rest, Input, Skip + 1, Stack, StringDecode, Value);
-terminate(<<>>, _Input, _Skip, _Stack, _StringDecode, Value) ->
+key_decode_function(Options) ->
+    case maps:find(keys, Options) of
+        error -> fun identity/1;
+        {ok, to_atom} -> fun to_atom/1;
+        {ok, to_existing_atom} -> fun to_existing_atom/1;
+        {ok, Invalid} -> error({invalid_option, #{keys => Invalid}})
```

Comment:
Let's either add this error case to `string_decode_function` or remove this case so it errors from the atom not being matched

Reply:
That makes sense. While refactoring, I had the thought we could use the same logic to decode both keys and strings (with separate options, of course). Something like this:

Reply:
I took the liberty of doing that change, so you can tell me what you think. If you'd rather have the old separate logic, let me know and I'll change it back 🙂

Reply:
I think it could be useful for some applications where you already have a subset of the data in code as atoms. E.g.:
```erlang
1> thoas:decode(Pokedex, #{keys => existing_atom, strings => existing_atom}).
{ok,#{pokemon =>
          [#{id => 1,name => <<"Bulbasaur">>,
             type => ['Grass','Poison'], % Atoms
             weight => <<"6.9 kg">>,
             weaknesses => ['Fire','Ice','Flying','Psychic'],  % Atoms
             spawn_time => <<"20:00">>,spawn_chance => 0.69,
             num => <<"001">>,
             next_evolution =>
                 [#{name => <<"Ivysaur">>,num => <<"002">>},
                  #{name => <<"Venusaur">>,num => <<"003">>}],
             multipliers => [1.58],
             img =>
                 <<"http://www.serebii.net/pokemongo/pokemon/001.png">>,
             height => <<"0.71 m">>,egg => <<"2 km">>,candy_count => 25,
             candy => <<"Bulbasaur Candy">>,avg_spawns => 69},
           ...]}}
```

Reply:
If you are still not convinced, I’ll remove it!

By the way, do you think the option `copy` has merit for keys?

Reply:
Thank you.

Ah yes, good point. Copying keys is probably useful for some people

Reply:
Should I leave it as is, or should we revert to a version where the errors are the same but with the old string decode logic and `copy` as an option for keys?

Reply:
Two functions please, so that values cannot be used for values.